MoreLinq/Interleave.cs

```diff
@@ -80,90 +52,41 @@ static IEnumerable<T> Interleave<T>(this IEnumerable<T> sequence, ImbalancedInte
+                var enumerators = sequences.Select(e => e.GetEnumerator()).Acquire();
 
                 try
                 {
-                    iteratorList = new List<IEnumerator<T>>(iterators);
-                    iterators = null;
-                    var shouldContinue = true;
-                    var consumedIterators = 0;
-                    var iterCount = iteratorList.Count;
-
-                    while (shouldContinue)
+                    var allNull = false;
+                    while (!allNull)
                     {
-                        // advance every iterator and verify a value exists to be yielded
-                        for (var index = 0; index < iterCount; index++)
+                        allNull = true;
+                        for (var i = 0; i < enumerators.Length; i++)
                         {
-                            if (!iteratorList[index].MoveNext())
-                            {
-                                // check if all iterators have been consumed and we can terminate
-                                // or if the imbalance strategy informs us that we MUST terminate
-                                if (++consumedIterators == iterCount || imbalanceStrategy == ImbalancedInterleaveStrategy.Stop)
-                                {
-                                    shouldContinue = false;
-                                    break;
-                                }
-
-                                iteratorList[index].Dispose(); // dispose the iterator sice we no longer need it
-
-                                // otherwise, apply the imbalance strategy
-                                switch (imbalanceStrategy)
-                                {
-                                    case ImbalancedInterleaveStrategy.Pad:
-                                        var newIter = iteratorList[index] = Generate(default(T), x => default).GetEnumerator();
-                                        newIter.MoveNext();
-                                        break;
-
-                                    case ImbalancedInterleaveStrategy.Skip:
-                                        iteratorList.RemoveAt(index); // no longer visit this particular iterator
-                                        --iterCount; // reduce the expected number of iterators to visit
-                                        --index; // decrement iterator index to compensate for index shifting
-                                        --consumedIterators; // decrement consumer iterator count to stay in balance
-                                        break;
-                                }
+                            var enumerator = enumerators[i];
+                            if (enumerator == null)
+                                continue;
 
+                            if (!enumerator.MoveNext())
+                            {
+                                enumerator.Dispose();
```

Comment:
Swap with next line.

Reply:
Fixed in 205c371a

Reply:
I think you misunderstood me here. I meant, instead of:

```c#
enumerator.Dispose();
enumerators[i] = null;
```

swap to this:

```c#
enumerators[i] = null;
enumerator.Dispose();
```

This is just good habit: update own state first then call external parties. If the enumerator's `Dispose` is well-implemented then it should never throw but if a poorly implemented one does then our state is consistent (that is, we consider the enumerator gone nonetheless) if someone catches and tries to resume iteration regardless.

Reply:
> This is just good habit: update own state first then call external parties.

Will made a tattoo of this one :)